src/lib.rs

```diff
@@ -6,34 +6,42 @@
-//! databases and blockchains. Byzantine fault tolerant systems can reach consensus with a number
-//! of faulty nodes _f_ (including complete takeover by an attacker), as long as the total number
-//! _N_ of nodes is greater than _3 f_.
-//!
-//! The Honey Badger consensus algorithm is both Byzantine fault tolerant and asynchronous. It does
-//! not make timing assumptions about message delivery. An adversary can control network scheduling
-//! and delay messages without impacting consensus, and progress can be made in adverse networking
-//! conditions.
+//! A consensus algorithm is a protocol that helps a number of nodes agree on some data value.
+//! Byzantine fault tolerant systems can tolerate a number of faulty nodes _f_ (broken, or even
+//! controlled by an attacker), as long as the total number _N_ of nodes is greater than _3 f_.
+//! Asynchronous protocols do not make assumptions about timing: Even if an adversary controls
+//! network scheduling and can delay message delivery, consensus will still be reached as long as
+//! all messages are _eventually_ delivered.
 //!
+//! The Honey Badger consensus algorithm is both Byzantine fault tolerant and asynchronous. It is
+//! also modular, and the subalgorithms it is composed of are exposed in this crate as well, and
+//! usable separately.
 //!
-//! ## Crate Implementation
+//! Consensus algorithms are fundamental to resilient, distributed systems such as decentralized
+//! databases and blockchains.
 //!
-//! This protocol does not function in a standalone context, it must be instantiated in an
-//! application that handles networking.
 //!
-//! * The network must contain a number of nodes that are known to each other by some unique
-//! identifiers (IDs) and are able to exchange authenticated (cryptographically signed) messages.
+//! ## Usage
 //!
-//! * The user must define a type of _input_ - the _transactions_ - to the system and nodes must
-//! handle system networking.
+//! This crate is meant to be used as a component in a distributed application where the consensus
+//! problem arises. The application will usually run on different nodes, connected to each other
+//! via a network. The nodes give an input to the algorithm, exchange several messages, and
+//! eventually the algorithm returns an output, which is guaranteed to be the same in each correct
+//! node.
 //!
-//! * Messages received from other nodes must be passed into the instance, and messages produced by
-//! the instance sent to corresponding nodes.
+//! However, the `hbbft` crate only implements the abstract protocols, not the networking. It is
+//! the application's responsibility to serialize, sign and send the messages to the other nodes.
+//! That is why in addition to methods for giving input, the algorithms also have a
+//! `handle_message` method. The application is required to call this for every (validly signed,
```

Comment:
call this method for every valid (signed and deserialized) message received from a peer.

Reply:
I'm a bit reluctant to refer to the whole message as "valid", since a faulty peer could have sent us a message with invalid content that has a valid signature and can be successfully deserialized. The validation of the message content happens inside our algorithms themselves, after the message has already been passed to `handle_message`.